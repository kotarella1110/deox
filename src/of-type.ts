import { filter } from 'rxjs/operators'

import { ActionCreator } from './create-action-creator'
import { AnyAction } from './create-action'
import { isOfType } from './is-of-type'
import { ExtractAction } from './types'

/**
 * Filter actions emitted by the source Observable by only emitting those that
 * are compatible with specified action(s) or action creator(s) or action type(s).
 *
 * @example
 * action$.pipe(
 *   ofType(foo),
 *   ...
 * )
 * @example
 * action$.pipe(
 *   ofType([foo, bar]),
 *   ...
 * )
 */
export function ofType<
  TSource extends AnyAction,
  TKey extends TSource['type'] | TSource | ActionCreator<TSource>,
  TSink extends TSource = ExtractAction<TKey, TSource>
>(keys: TKey | ReadonlyArray<TKey>) {
  return filter<TSource, TSink>(isOfType(keys))
}
