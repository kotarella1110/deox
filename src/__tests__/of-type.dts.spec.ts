import { of } from 'rxjs'

import { createActionCreator } from '../create-action-creator'
import { ofType } from '../of-type'

// @dts-jest:group ofType

const a = createActionCreator('a')
const b = createActionCreator('b')
const c = createActionCreator('c')
const d = createActionCreator('d')

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType(a))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType([a, b]))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType(a()))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType([a(), b()]))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType('a'))

// @dts-jest:pass:snap
of(a(), b(), c()).pipe(ofType(<const>['a', 'b']))

// @dts-jest:pass:snap
of(a(), b(), c(), d()).pipe(ofType(<const>['a', b(), c]))
