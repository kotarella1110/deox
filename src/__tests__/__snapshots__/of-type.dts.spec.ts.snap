// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ofType of(a(), b(), c()).pipe(ofType('a')) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType(<const>['a', 'b'])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType([a(), b()])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType([a, b])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType(a())) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of(a(), b(), c()).pipe(ofType(a)) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; }>"`;

exports[`ofType of(a(), b(), c(), d()).pipe(ofType(<const>['a', b(), c])) (type) should match snapshot 1`] = `"Observable<{ type: \\"a\\"; } | { type: \\"b\\"; } | { type: \\"c\\"; }>"`;
